Storage/SQL/SQLiteHistoryRecommendations.swift

```diff
@@ -171,43 +180,42 @@ extension SQLiteHistory: HistoryRecommendations {
         return [(clearHighlightsQuery, nil), (sql, args)]
     }
 
-    // Checks if there are more than 100k items in the `visits` table.
-    // This is used as an indicator that we should clean up old history.
-    func checkIfCleanupIsNeeded() -> Deferred<Maybe<Bool>> {
-        let sql = "SELECT COUNT(*) FROM \(TableVisits)"
+    // Checks if there are more than the specified number of rows in the
+    // `history` table. This is used as an indicator that `cleanupOldHistory()`
+    // needs to run.
+    func checkIfCleanupIsNeeded(maxHistoryRows: UInt) -> Deferred<Maybe<Bool>> {
+        let sql = "SELECT CASE WHEN COUNT(rowid) > \(maxHistoryRows) THEN 1 ELSE 0 END AS cleanup FROM \(TableHistory)"
         return self.db.runQuery(sql, args: nil, factory: IntFactory) >>== { cursor in
-            guard let visitCount = cursor[0], visitCount > 100000 else {
+            guard let cleanup = cursor[0], cleanup > 0 else {
                 return deferMaybe(false)
             }
 
             return deferMaybe(true)
         }
     }
 
-    // Deletes the oldest 50k items from the `visits` table and their
-    // corresponding items in the `history` table. This only gets run when
-    // there are more than 100k items in the `visits` table. Because of
-    // this, we may need to clean up several times until we've crossed
-    // below the 100k-item threshold.
-    private func cleanupOldHistory() -> [(String, Args?)] {
-        let visitsSQL = """
-            DELETE FROM \(TableVisits) WHERE id IN (
-                SELECT id FROM \(TableVisits) ORDER BY \(TableVisits).date ASC LIMIT 50000
-            )
-            """
-        let historySQL = """
-            DELETE FROM \(TableHistory) WHERE NOT EXISTS (
-                SELECT 1 FROM \(TableVisits) WHERE \(TableVisits).siteID = \(TableHistory).id
+    // Deletes the specified number of items from the `history` table and
+    // their corresponding items in the `visits` table. This only gets run
+    // when the `checkIfCleanupIsNeeded()` method returns `true`. It is possible
+    // that a single clean-up operation may not remove enough rows to drop below
+    // the threshold used in `checkIfCleanupIsNeeded()` and therefore, this may
+    // end up running several times until that threshold is crossed.
+    func cleanupOldHistory(numberOfRowsToPrune: UInt) -> [(String, Args?)] {
+        let sql = """
+            DELETE FROM \(TableHistory) WHERE id IN (
+                SELECT id FROM \(TableHistory)
+                ORDER BY CASE WHEN local_modified THEN local_modified ELSE server_modified END ASC
```

Comment:
If `local_modified` is either null or not null, you can do:

```
ORDER BY ifnull(local_modified, server_modified)
```

If you want to take the largest (and I think you do!), use `max` and `ifnull` together:

```sql
sqlite> select * from foo;
a           x           y
----------  ----------  ----------
one         5
two                     6
three       2           10
sqlite> select a, ifnull(x, y), max(x, y) as naive, max(ifnull(x, 0), ifnull(y, 0)) as realmax from foo order by realmax;
a           ifnull(x, y)  naive       realmax
----------  ------------  ----------  ----------
one         5                         5
two         6                         6
three       2             10          10
```

so here:

```sql
ORDER BY max(ifnull(local_modified, 0), ifnull(server_modified, 0)) ASC
```

Reply:
Thank you. I knew there was a way to do this, but I couldn't remember what it was.